src/cmap/connection_pool.ts

```diff
@@ -354,17 +355,33 @@ export class ConnectionPool extends TypedEventEmitter<ConnectionPoolEvents> {
    * will be held by the pool. This means that if a connection is checked out it MUST be checked back in or
    * explicitly destroyed by the new owner.
    */
-  async checkOut(): Promise<Connection> {
+  async checkOut(options?: { timeout?: Timeout }): Promise<Connection> {
     this.emitAndLog(
       ConnectionPool.CONNECTION_CHECK_OUT_STARTED,
       new ConnectionCheckOutStartedEvent(this)
     );
 
     const waitQueueTimeoutMS = this.options.waitQueueTimeoutMS;
+    const serverSelectionTimeoutMS = this[kServer].topology.s.serverSelectionTimeoutMS;
 
     const { promise, resolve, reject } = promiseWithResolvers<Connection>();
 
-    const timeout = Timeout.expires(waitQueueTimeoutMS);
+    let timeout: Timeout;
+    if (options?.timeout) {
+      // CSOT enabled
+      // Determine if we're using the timeout passed in or a new timeout
+      if (
+        csotMin(options.timeout.duration, serverSelectionTimeoutMS) === serverSelectionTimeoutMS
+      ) {
+        timeout = Timeout.expires(serverSelectionTimeoutMS - options.timeout.timeElapsed);
+      } else {
+        timeout = options.timeout;
+      }
+    } else {
+      timeout = Timeout.expires(waitQueueTimeoutMS);
+    }
+
+    timeout.throwIfExpired();
```

Comment:
Good call. Do we want to encapsulate the throwIfExpired calls into a `Timeout.race<T>(value: Promise<T>)` method? it would be pretty much just call throwIfExpired and then do the same Promise.race call as we do currently.

That would make it such that we only call throwIfExpired just as we go to complete the blocking work, which would guarantee that we catch the timeout error whenever it occurs.

Reply:
On paper, sounds good, in practice, if there are any instances where we need to not even begin work if this is already expired that is hard to accomplish with a helper that isn't aware of the context you're in.

Reply:
That's a good point about needing the context. As far as the question about whether or not `checkout started` is correct if we're given an expired timeout. We're considering the case here that server selection succeeded, some amount of time passed between server selection succeeding and entering `ConnectionPool.checkOut` and the timeout expiring in that intermediate time.

I can see a strong argument for emitting the checkout started event as we enter `checkOut` (which we do currently), but not putting the waitqueue member in the queue at all if the timeout has already expired so as not to waste work. As you point out, the only time this would occur is when CSOT is enabled and the configured `serverSelectionTimeoutMS` is greater than the configured `timeoutMS`, meaning we'd be using the timeout passed in on the options. In that case, it would be safest and make the most sense to throw a `MongoOperationTimeoutError` with the timeout being the cause, but I'm not sure how that gels with existing expectations for the CMAP spec.

If we just want to proceed with this PR, I'd argue that we should call throwIfExpired just before the call to Promise.race (after we put the waitqueue member in the queue) since that's the beginning of the blocking work which we'd then cancel and that's what the CSOT spec is primarily concerned with and we can file a ticket to address this if it ends up causing further issues.